### DISCUSSION TITLE
[need help about individual Elemental value] A weird point about interior structure of moose

### DISCUSSION THREAD

**Discussion by Ttw0626** [idaholab/moose#27983] [Q&A Modules: Solid mechanics] [answered]
Hi, moose-developers,

**Question description**:
after a lot of attempts by myself, I cannot understand 
**How to handle with the elemental communication between Material and UserObject ?**

- **_evidence 1_**: 
Try ```AbaqusUserElement```  UO  with external ```uel.f```
**[which only output current element ID with jelem=elem->id()+1]**,
the output on terminal as shown in below:
  - mesh information: 
    >                                Parallelism:
    >                                         Num Processors:            4
    >                                         Num Threads:               1
    >                                Mesh:
    >                                         Parallel Type:             replicated
    >                                         Mesh Dimension:            3
    >                                         Spatial Dimension:         3
    >                                Nodes:
    >                                         Total:                     64
    >                                         Local:                     30
    >                                         Min/Max/Avg:               6/30/16
    >                                Elems:
    >                                         Total:                     27
    >                                         Local:                     8
    >                                         Min/Max/Avg:               6/8/6
    >                                         Num Subdomains:            1
    >                                         Num Partitions:            4
    >                                         Partitioner:               metis
  - Output [1. the execute_on ```EXEC_PRE_KERNELS```; 2. each processor outputs all elements totally 27 elements]:
    >                                 Time Step 0, time = 0
    >
    >                                 Time Step 1, time = 10, dt = 10
    >                                 Element number is:            1
    >                                 Element number is:           13
    >                                 Element number is:           14
    >                                 Element number is:           15
    >                                 Element number is:           17
    >                                 Element number is:            2
    >                                 Element number is:            3
    >                                 Element number is:           10
    >                                 Element number is:           18
    >                                 Element number is:            4
    >                                 Element number is:           16
    >                                 Element number is:           11
    >                                 Element number is:           22
    >                                 Element number is:           24
    >                                 Element number is:            5
    >                                 Element number is:           12
    >                                 Element number is:            6
    >                                 Element number is:           23
    >                                 Element number is:           26
    >                                 Element number is:           19
    >                                 Element number is:           27
    >                                 Element number is:            7
    >                                 Element number is:           25
    >                                 Element number is:           20
    >                                 Element number is:            8
    >                                 Element number is:           21
    >                                 Element number is:            9
    >
    >                                 Element number is:           13
    >                                 Element number is:           15
    >                                 Element number is:            1
    >                                 Element number is:            2
    >                                 Element number is:           10
    >                                 Element number is:           14
    >                                 Element number is:           17
    >                                 Element number is:            3
    >                                 Element number is:           18
    >                                 Element number is:           11
    >                                 Element number is:           16
    >                                 Element number is:            4
    >                                 Element number is:            5
    >                                 Element number is:           22
    >                                 Element number is:           24
    >                                 Element number is:           12
    >                                 Element number is:            6
    >                                 Element number is:           23
    >                                 Element number is:           26
    >                                 Element number is:           19
    >                                 Element number is:            7
    >                                 Element number is:           20
    >                                 Element number is:           27
    >                                 Element number is:           25
    >                                 Element number is:            8
    >                                 Element number is:           21
    >                                 Element number is:            9
    >
    >                                 Element number is:            1
    >                                 Element number is:           13
    >                                 Element number is:           15
    >                                 Element number is:            2
    >                                 Element number is:           17
    >                                 Element number is:           10
    >                                 Element number is:           14
    >                                 Element number is:            3
    >                                 Element number is:           16
    >                                 Element number is:           18
    >                                 Element number is:           11
    >                                 Element number is:            4
    >                                 Element number is:           12
    >                                 Element number is:           22
    >                                 Element number is:           24
    >                                 Element number is:            5
    >                                 Element number is:           26
    >                                 Element number is:           19
    >                                 Element number is:           23
    >                                 Element number is:            6
    >                                 Element number is:           25
    >                                 Element number is:           27
    >                                 Element number is:           20
    >                                 Element number is:            7
    >                                 Element number is:           21
    >                                 Element number is:            8
    >                                 Element number is:            9
    >
    >                                 Element number is:            1
    >                                 Element number is:           13
    >                                 Element number is:           15
    >                                 Element number is:            2
    >                                 Element number is:           17
    >                                 Element number is:           10
    >                                 Element number is:            3
    >                                 Element number is:           14
    >                                 Element number is:           11
    >                                 Element number is:           18
    >                                 Element number is:            4
    >                                 Element number is:           16
    >                                 Element number is:            5
    >                                 Element number is:           12
    >                                 Element number is:           24
    >                                 Element number is:           22
    >                                 Element number is:            6
    >                                 Element number is:           19
    >                                 Element number is:           26
    >                                 Element number is:           23
    >                                 Element number is:            7
    >                                 Element number is:           20
    >                                 Element number is:           27
    >                                 Element number is:           25
    >                                 Element number is:            8
    >                                 Element number is:           21
    >                                 Element number is:            9
  - Conclusion: 
    I do not know it's expected result of the abaqus user-element implementation in moose or not. However, it seems like each processor called the plugin ```uel.f``` across all 27 elements, but we clearly see the correct parallel partition with 4 cpus.
  - **Question 1**:
    From the above background about ```AbaqusUserElement```, If I deleted the ```DynamicLibraryLoader``` function especially for ```uel.f``` Plugin, I wanna directly output the ```jelem=elem->id()+1``` within ```UELThread.C``` w/o plugin function ```_uel()```:
    >                       std::cout << "Element number is           " << jelem << "\n";
    the new wrong Output is not the same as the above old Output with plugin ```uel.f```, I cannot understand this difference, how can I get the same Output without plugin ?
    **new wrong Output**:
    >                                 Time Step 0, time = 0
    >
    >                                 Time Step 1, time = 10, dt = 10
    >                                 Element number is:            1
    >                                 Element number is:           13
    >                                 Element number is:           14
    >                                 Element number is:           15
    >                                 Element number is:           17
    >                                 Element number is:            2
    >                                 Element number is:            3
    >                                 Element number is:           10
    The new wrong Output only for one processor and also executed on ```EXEC_PRE_KERNELS```.

- _**evidence 2**:_
Try sub- UO class based on parent class ```ElementUserObject```  
**[which only output current element ID with _current_elem->id()+1]**,
the output on terminal as shown in below:
  - mesh information [same]: 
    >                                Parallelism:
    >                                         Num Processors:            4
    >                                         Num Threads:               1
    >                                Mesh:
    >                                         Parallel Type:             replicated
    >                                         Mesh Dimension:            3
    >                                         Spatial Dimension:         3
    >                                Nodes:
    >                                         Total:                     64
    >                                         Local:                     30
    >                                         Min/Max/Avg:               6/30/16
    >                                Elems:
    >                                         Total:                     27
    >                                         Local:                     8
    >                                         Min/Max/Avg:               6/8/6
    >                                         Num Subdomains:            1
    >                                         Num Partitions:            4
    >                                         Partitioner:               metis
  - Output [1. the execute_on ```EXEC_PRE_KERNELS```]:
    >                                 Time Step 0, time = 0
    >
    >                                 Time Step 1, time = 10, dt = 10
    >                                 Element number is:           10
  - Conclusion: 
    this output is totally incorrect. From my view, I just wanna output the individual ElementID of each element. This is to say, the ```threadJoin( )``` and ```finalize( )``` are not necessary for this expectation. In ```AbaqusUserElement``` UO, it did not use the the ```threadJoin( )``` and ```finalize( )``` as well.

- **Question**:
    From the two evidence 1, 2, The only ```AbaqusUserElement``` can output ElementID of all elements but duplicated.
    Although this ElementID example is simple, it's important to implement elemental value individually. **Is there any way to focus on only each element ?** by the way, I understood the UOs ```BlockAverageValue``` and ```HomogenizationConstraints```, which calculated Elemental value from quadrature points loop, threadJoining into shared map container, and gatherSum to global value. However, it is different from the expectation of this discussion, I just wanna perform the quadrature points loop, calculate each Elemental value, and output it element by element. This idea maybe the same as ```Elemental value postprocessor``` for ```Paraview``` visualization.

    the descent trend : 
    
    **domain (FEProblem)** (global)   --->  **subProblems** (processors)  --->  **subDomains (blocks)** (each processor)
    
    **Elements (block)** (each processor)  --->  **quadrature points (Element)** (each processor)  --->
   
    **quadrature point** (each processor), is my understanding correct ?

    **Maybe I can try output each ElementID within** ```computeProperties()``` **, the result could be correct ?**

Thanks, if developers need any info., plz feel free let me know.
best regards,
tianwen

**Comment by GiudGiud** [comment]
Hello

> How to handle with the elemental communication between Material and UserObject ?

There are two directions here.
From material to UO:
- materials precompute material properties, which UOs can just retrieve a reference to with getMaterialProperty

From UO to material:
- materials can retrieve pointers to user objects of any type using the templated getUserObject<> so they can call any public API on those

For the AbaqusElementUO questions I dont know. @dschwen

**Reply by Ttw0626** [reply]
hi, giud, I understood your kind answers. Thanks.
I think I did not explain my question very well.

- these operations make sense to me, we can get the Material properties and the data calculated within UOs
   via ```getMaterialProperty<>()``` and ```getUserObject<>()```. The example ```BlockAverageValue``` in moose
   explained it nicely.

- I wanna dicuss Two situations, how to get expected output ? :
   - If I wanna output individually the ```ElementID``` of each element, could I just need to do this I/O operation within
      ```computeProperties()``` ? but I confused that "the computeProperties() is work on element or (subdomains) processor ?"
   
   - although it's duplicated job, I wanna store one MaterialProperty<Real>[_qp] of each quadrature point within one shared
      map container (```std::map<dof_id_type, std::vector<Real>>   map[elementID][quadraturePoints]```) via UO, and then, to get this
      materialProperty of each quadrature point at Materials via 
      (```Real thisMaterialProperty = UO.getMap(_current_element->id())[_qp]```).

From the first case, the expected output should be ```Element Number is    x``` looped over each element and output this string number of elements times on terminal.

From the second case, the output of materialProperty from UO stored and the output of materialProperty by directly ```declareMaterialProperty``` in ```computeQpProperties()``` should be the same, but I tried before, it's not same. There is not any
communication between element by element and processor by processor, so I did not use ```threadJoin()``` and ```finalize()```.

If I got any misunderstanding, plz tell me ! Thanks

**Reply by Ttw0626** [reply]
Updated...........

I tried directly output ElementID via ```computeProperties()``` or ```computeQpProperties()``` if and only if ```_tid=0 && _qp=0``` condition.
The output of thread ID = 0 (maybe master) presented including 0th partial outputs of ElementID, two case are the same.

dose the terminal output have a constraint only for threadID 0 ? only threadID 0 processor will output on terminal.
how can I output ElementID of all elements ?

could I explain it if I did not do ```threadJoin( )``` and ```finalize( )``` within UO, the only threadID0 read the data of UO at materials ?

**Reply by GiudGiud** [reply]
use `--keep-cout` on the command line to keep all the output.
I dont think we stop output from other threads, but we do stop output from other mpi processes.

**Reply by GiudGiud** [reply]
> f I wanna output individually the ElementID of each element, could I just need to do this I/O operation within
computeProperties() ? but I confused that "the computeProperties() is work on element or (subdomains) processor ?"

for what purpose? To look at it in a CSV?
Use an Element...VPP for that. They will output the id and locations

> although it's duplicated job, I wanna store one MaterialProperty[_qp] of each quadrature point within one shared
map container (std::map<dof_id_type, std::vector<Real>>   map[elementID][quadraturePoints]) via UO, and then, to get this
materialProperty of each quadrature point at Materials via
(Real thisMaterialProperty = UO.getMap(_current_element->id())[_qp]).

using a map for material properties is unconventional and not really necessary. You'll need to be careful to execute the User Object early enough that the map is ready for wherever you plan to use it


Are you using threads? Or MPI? Or both?
For forming a map of all properties, I expect you may need to use `threadJoin()` for threads to perform a reduction of that map, though unless you want to distribute the map, I expect you do not need to use `finalize()`

**Reply by Ttw0626** [reply]
> use `--keep-cout` on the command line to keep all the output. I dont think we stop output from other threads, but we do stop output from other mpi processes.

hi giud, this make sense to me, Thanks.

**Reply by Ttw0626** [reply]
> > f I wanna output individually the ElementID of each element, could I just need to do this I/O operation within
> > computeProperties() ? but I confused that "the computeProperties() is work on element or (subdomains) processor ?"
> 
> for what purpose? To look at it in a CSV? Use an Element...VPP for that. They will output the id and locations
> 
> > although it's duplicated job, I wanna store one MaterialProperty[_qp] of each quadrature point within one shared
> > map container (std::map<dof_id_type, std::vector>   map[elementID][quadraturePoints]) via UO, and then, to get this
> > materialProperty of each quadrature point at Materials via
> > (Real thisMaterialProperty = UO.getMap(_current_element->id())[_qp]).
> 
> using a map for material properties is unconventional and not really necessary. You'll need to be careful to execute the User Object early enough that the map is ready for wherever you plan to use it
> 
> Are you using threads? Or MPI? Or both? For forming a map of all properties, I expect you may need to use `threadJoin()` for threads to perform a reduction of that map, though unless you want to distribute the map, I expect you do not need to use `finalize()`


- **(1)** it's just simple test for my purpose, NVM. I will explain my purpose in below. sorry for this confusion.

- **(2)** What I wanna do is, for instance, we consider HEX8 lagrangian element with 8 quadrature points
  - [1] calculate one scalar material property at quadrature point [_qp] in ```ComputeQpProperties()```, such as hydrostatic stress. this done at time increment n.
    
    -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  - [2] extra this hydrostatic stress material property < Real > and store them into a **mapStored** within UO by **mapStored**[elementIDs][quadrature points], as giud you said using ```getMaterialProperty<>```
    for instance, this material property at 0th quadrature point within 4th element, which stored in **mapStored**[4][0].
    
  - [3] do some elemental operations with 8 quadrature points from above stored **mapStored**, after that, we calculated a new material property < RealVectorValue > at each quadrature point. store these new material properties into **mapNew** [elementIDs][quadrature points], which contains < RealVectorValue >.

    -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
    
  - [4] we focus on time increment n+1 and go back to Material System. To obtain this **mapNew** in ```ComputeQpProperties()``` as giud you said using ```getUserObject<>```. this new calculated < RealVectorValue > by UO will be used to calculate another MaterialProperty in current timestep, such as flux, and provide to ```computeQpResidual``` and ```computeQpJacobian```.


This is what I ask the question about individual Elemental value, all UO operations just use quadrature points of same elements like tiny User-Element.
I already tried some followed above procedure, The result is strongly changed with different processors (MPI) used.
I did not use threads only MPI, is that means I do not need to use ```threadJoin( )``` and ```finalize( )``` ?

regards,
tianwen

**Reply by GiudGiud** [reply, marked as answer]
To use threads you must implement threadJoin whenever a reduction operation is needed.

To use MPI / processes, similar for finalize()

> The result is strongly changed with different processors (MPI) used

The element numbering is different in parallel. Use `allow_renumbering = false` in the Mesh block to keep the same ordering

> [3] do some elemental operations with 8 quadrature points from above stored mapStored, after that, we calculated a new material property < RealVectorValue > at each quadrature point. store these new material properties into mapNew [elementIDs][quadrature points], which contains < RealVectorValue >.

I would recommend creating a second material property for the property retrieved by a UO then, this will simplify the logic.
I think you have the workflow well set up already?
Did you check that the user object is expected before the kernels/bcs/etc that need this custom material property?

**Reply by Ttw0626** [reply]
Hi, giud, I re-view the Moose workshop slides and tried some different UOs-based.

for now, I got the correct results. the **KeyPoint** from my view, we should maintain the same data storage ```map``` over all processors, this is my previous misunderstanding. the replicated mesh as default.

Thanks for your all suggestions and help. I will mark as answer.

best regards,
tianwen
